_posts/2024-02-09-Database-interaction-with-skunk-scala-library:-A-comprehensive-guide.md

```diff
@@ -0,0 +1,555 @@
+    ) derives ConfigReader
+```
+
+## 5. Resources
+Configuration values reflecting our domain to be used throughout our project.
+
+* `application.conf` placed in `src/main/resources`. Here we set our configuration values needed to establish our database session. The key-value pairs that are set up in this file must correspond to the values specified in our `Config` case class.
+```scala
+    host: "localhost"
+    port: 5432
+    username: "postgres"
+    password: "postgres"
+    database: "skunk"
+```
+* `tables.sql` placed in `src/main/resources`. We need to write a SQL statement to create a `users` table in our database. As previously mentioned, the **fields** and **data types** of the created table (`users` in our case) should match the **values** and **data types** of the domain to be stored in (`User` in our case).
+```sql
+    CREATE TABLE IF NOT EXISTS users (
+        id      UUID    PRIMARY KEY,
+        name    VARCHAR        NOT NULL,
+        email   VARCHAR UNIQUE NOT NULL
+    );
+```
+* We can now create a `docker-compose.yml` file (which can be placed in the root directory of our project) to spin up a docker container running a postgreSQL image with the given environment values. Please ensure that the database credentials in this file match the postgres configuration values set in `application.conf`. `tables.sql` is used by docker as an entry point to initialize our database and provide a volume to store our data.
+```scala
+    version: "3.3"
+
+    services:
+      postgres:
+        restart: always
+        image: postgres:13.0-alpine
+        ports:
+          - "5432:5432"
+        environment:
+          - DEBUG=false
+          - POSTGRES_USER=postgres
+          - POSTGRES_PASSWORD=postgres
+          - POSTGRES_DB=skunk
+        volumes:
+          - ./src/main/resources/tables.sql:/docker-entrypoint-initdb.d/init.sql
+```
+
+## 6. Setting Up Database Connection
+
+To begin interacting with the database using Skunk, establish a database connection. This involves configuring the connection parameters such as host, port, username, password, and database name. Skunk provides a convenient way to manage database sessions using resources.
+
+Before establishing our database connection, let's start our Docker container by running the following command; `docker-compose up` in the terminal at the root directory or wherever we placed our `docker-compose.yml` file.
+
+We then create a `main.scala` file in our project where we will test our codes.
+Good! Now let's create a class named `DbConnection` to establish a database connection, allowing us to choose between two connection types: `single` or `pooled`, based on the provided configuration values.
+```scala
+    import cats.effect.Temporal
+    import cats.effect.std.Console
+    import fs2.io.net.Network
+    import natchez.Trace
+    import skunk.Session
+    import skunk_guide.domain.Config
+
+    final class DbConnection[F[_] : Temporal : Trace : Network : Console] {
```

Comment:
Let's make this an object, and put the `[F[_] : Temporal : Trace : Network : Console]` in both methods